src/models/CommonModel.ts

```diff
@@ -20,6 +20,102 @@ export class CommonModel extends CommonSchema<CommonModel> {
   extend?: string[];
   originalSchema?: Schema | boolean;
   
+  /**
+   * Retrieves data from originalSchema by given key
+   * 
+   * @param key given key
+   * @returns {any}
+   */
+  getFromSchema<K extends keyof Schema>(key: K) {
+    let schema = this.originalSchema || {};
+    if (typeof schema === 'boolean') schema = {};
+    return schema[`${key}`];
+  }
+
+  /**
+   * Set the types of the model
+   * 
+   * @param types to set the model type to
+   */
+  setType(types : string | string[]) {
+    if (Array.isArray(types)) {
+      if (types.length === 0) {
+        this.type = undefined;
+        return;
+      } else if (types.length === 1) {
+        this.type = types[0];
+        return;
+      }
+    }
+    this.type = types;
+  }
+
+  /**
+   * Adds types to the existing model types.
+   * 
+   * Makes sure to only keep a single type incase of duplicates.
+   * 
+   * @param types which types we should try and add to the existing output
+   */
+  addToTypes(types: string[] | string) {
+    if (Array.isArray(types)) {
+      types.forEach((value) => {
+        this.addToTypes(value);
+      });
+    } else if (this.type === undefined) {
+      this.type = types;
+    } else if (!Array.isArray(this.type) && this.type !== types) {
+      this.type = [this.type, types];
+    } else if (Array.isArray(this.type) && !this.type.includes(types)) {
+      this.type.push(types);
+    }
+  }
+
+  /**
+   * Checks if given property name is required in object
+   * 
+   * @param propertyName given property name
+   * @returns {boolean}
+   */
+  isRequired(propertyName: string): boolean {
+    if (this.required === undefined) {
+      return false;
+    }
+    return this.required.includes(propertyName);
+  }
+
+  /**
+   * This function returns an array of `$id`s from all the CommonModel's it immediate depends on.
+   */
+  getImmediateDependencies(): string[] {
+    const dependsOn = [];
+    if (this.additionalProperties instanceof CommonModel) {
+      const additionalPropertiesRef = (this.additionalProperties as CommonModel).$ref;
+      if (additionalPropertiesRef !== undefined) {
+        dependsOn.push(additionalPropertiesRef);
+      }
+    }
+    if (this.extend !== undefined) {
+      for (const extendedSchema of this.extend) {
+        dependsOn.push(extendedSchema);
+      }
+    }
+    if (this.items instanceof CommonModel) {
+      const itemsRef = (this.items as CommonModel).$ref;
+      if (itemsRef !== undefined) {
+        dependsOn.push(itemsRef);
+      }
+    }
+    if (this.properties !== undefined && Object.keys(this.properties).length) {
+      Object.entries(this.properties).forEach(([, propertyModel]) => {
```

Comment:
I guess you could just use `Object.values` instead of `Object.entries`. Also imo it is more readable if you instead of `forEach` use `map`, and then push it with spread like `dependsOne.push(...result)`.

Reply:
I agree with `Object.values` is better 👍 

Not sure about map since we have to filter it 🤔 Guess it comes down to preference. Your suggested approach can be:

```ts
    if (this.properties !== undefined && Object.keys(this.properties).length) {
      const referencedProperties = Object.values(this.properties)
        .filter((propertyModel) => {return propertyModel.$ref !== undefined;})
        .map((propertyModel) => {return `${propertyModel.$ref}`;});
      dependsOn.push(...referencedProperties);
    }
```

Do you think this is better? 🤔

Reply:
```
if (this.properties !== undefined && Object.keys(this.properties).length) { 
const referencedProperties = Object.values(this.properties)
.filter(propertyModel =>  propertyModel.$ref !== undefined) // you do not have to open braces in arrow function if you just return
.map(propertyModel => propertyModel.$ref); // is not $ref already a string
dependsOn.push(...referencedProperties); 
}
```

Yes I do, because you do not mutate an object inside a scope. Now if you move this part to a separete function
```
Object.values(this.properties)
.filter(propertyModel =>  propertyModel.$ref !== undefined) // you do not have to open braces in arrow function if you just return
.map(propertyModel => propertyModel.$ref);
```
unit test is easier to make because you do not modify internal state of class inside of this function.

Reply:
True, need to remember JS have multiple ways to simplify the syntax 😄

Reply:
Personally I would write something like this.
```
const getReferences = (properties) => {
if(properties === undefined) {
return [];
}
const referenceNames = Object.values(properties);
if(referenceNames.length === 0) {
return [];
}
return referenceNames 
.filter(propertyModel =>  propertyModel.$ref !== undefined)
.map(propertyModel => propertyModel.$ref)
}
```
```
dependsOn.push(...getReferences(this.properties));
```
Now you can make unit test which cover more cases. Also it will reduce size of function where getReferences is called.

Reply:
One thing we need to be aware of is that this causes the array to be iterated twice when there is no need for it. In our case it does not affect it that much, however, it all depends on how large your dataset is... [Might be time for an eslint rule for loops](https://github.com/github/eslint-plugin-github/blob/main/docs/rules/array-foreach.md#chaining)?

Reply:
@jonaslagoni 
ok, reduce would be better
```ts
return propertyModels
     .reduce((acc, propertyModel) => 
        propertyModel.$ref !== undefined
            ? [...acc, propertyModel.$ref]
            : acc
     ,[]);
```

Reply:
Also be aware that a lot of eslint rules was written for earlier versions of node.js.

As example in node 10 they changed how promise trace is logged, so using `return await Promise.resolve()` is better than `return Promise.resolve()`. Previously trace was the same but returning without await is a microoptimalization, now trace is showed up to the point where you use `await` not farther.

Reply:
I also like this, but if I have to look at it with criticism and a new pair of eyes.

As soon as we start to introduce chaining, we simply reduce the clarity of the function. Newer developers might not directly know `reduce` increasing the complexity.

And since it is open-source we should as much as possible, have as low complexity as possible. I have a hard time finding the balance my self also why we inconsistently use different for loops.

```ts
Object.values(this.properties).forEach((propertyModel) => { 
  if (propertyModel.$ref !== undefined) {
      dependsOn.push(propertyModel.$ref);
  }
});
```
right of the bat everyone can probably guess what `Object.values` does, and so with `forEach`.

With filter and map chaining it becomes somewhat clearer.
```ts
const referencedProperties = Object.values(this.properties)
  .filter(propertyModel => propertyModel.$ref !== undefined)
  .map(propertyModel => `${propertyModel.$ref}`);
dependsOn.push(...referencedProperties);
```

With reduce it becomes somewhat more complex to understand what is happening: 
```ts
propertyModels
  .reduce((acc, propertyModel) => 
    propertyModel.$ref !== undefined
      ? [...acc, propertyModel.$ref]
      : acc
  ,[]);
```

I suggest we should stick with `filter` and `map` for now and simply accept the speed penalty. What do you think @czlowiek488 ? 

Gonna create an issue so we can agree on a consistent way of iterating arrays in the library, also in terms of `forEach`, `for of` etc.

Reply:
https://github.com/asyncapi/generator-model-sdk/issues/192

Reply:
> As soon as we start to introduce chaining, we simply reduce the clarity of the function.
I do not think that chaining reduce clarity, I can agree that it reduce maintability.
> Newer developers might not directly know reduce increasing the complexity.
I totally disagree, if someone do not know the language it is his problem. Also this is in every JS documentation I know about so there is no reason for him to read docs and get this knowledge. So if developer do not know what reduce do, it was his decision for not reading specs. We should use all JS newest features to make possibly fastest and cleanest code.
> And since it is open-source we should as much as possible, have as low complexity as possible. I have a hard time finding the balance my self also why we inconsistently use different for loops.
When I would see filter and map (or something else) instead of reduce I have a question in my head "For what reasons did they not use reduce?" first thoughs is "Because reduce does not allow to do it" but... reduce can do literally everything with array. So I would investigate why it was not used and then if I would found that it could be used... it would give me a headache and make me swear the developer who wrote this.

Reply:
Another reason to use reduce is that you need to make more units to test if you would like.
So as example.
```ts
// 2 units for testing
export const filterProperties = propertyModel => propertyModel.$ref !== undefined;
export const mapProperties = propertyModel =>propertyModel.$ref;
Object.values(this.properties)
  .filter(filterProperties)
  .map(mapProperties)
```
vs
```ts
// 1 unit for testing
export const reducePropertes = (acc, propertyModel) => 
    propertyModel.$ref !== undefined
      ? [...acc, propertyModel.$ref]
      : acc;
Object.values(this.properties)
  .reduce(reduceProperties,[]);
```
Sorry for variable names, I do not want to think about it in example.

Reply:
> I totally disagree, if someone do not know the language it is his problem. Also this is in every JS documentation I know about so there is no reason for him to read docs and get this knowledge. So if developer do not know what reduce do, it was his decision for not reading specs. We should use all JS newest features to make possibly fastest and cleanest code.

Hehe, yea I can easily agree that in most cases you would want to use the newest features, but keep in mind that we want to make time to contribution as low as possible. If this means not using a newer feature for it to be more maintainable so be it, but it all depends. It can vary when it makes sense to prioritize what.

> Another reason to use reduce is that you need to make more units to test if you would not.
I disagree here, yes in your scenario you have two units you are testing, but for this function, you have `getImmediateDependencies`. Nothing more nothing less, and the tests you are creating should not directly care whether you use `filter+map` or `reduce`, what you are testing is the same, especially in terms of coverage.

Also, we cannot use `reduce` in TS it restricts the use case and the return type to `CommonModel` not an array as you use it here. While it is doable in pure js, in TS it is not 🙂 

So lets stay with `filter+map`.

Reply:
@jonaslagoni  You can type reduce easly.

Reply:
> @jonaslagoni You can type reduce easly.

We might be able to do so, but let's not push the use-case of the function, filter+map will do the job nicely in my opinion.